### PR TITLE
Progressive Point Cloud Deconvolution Generation Network

### PR DESCRIPTION
**Pull request by CCInc**
https://arxiv.org/pdf/2007.05361.pdf

It is fully implemented and will run but I am unable to train it for long enough. In the paper they said it trained for 1000 epoch over 2 days.

### CODE REVIEW

**Comment by CCInc** [comment]
`python3 train.py task=generative model_type=dgcn model_name=PDGN_Paper dataset=shapenet training=dgcn training.batch_size=16`

File: conf/models/generative/dgcn.yaml [hunk 1]
```diff
@@ -0,0 +1,27 @@
+models:
+    PDGN_Paper:
+        class: DGCN.DGCN
+        conv_type: "SPARSE"
+        scales: [256, 512, 1024, 2048] # number of progressive points the generator generates
```
Comment (nicolas-chaulet) [inline conf/models/generative/dgcn.yaml:5]: You can use ${data.scales}

File: conf/visualization/default.yaml [hunk 2]
```diff
@@ -1,9 +1,7 @@
 visualization:
-  activate: False
+  activate: True
```
Comment (nicolas-chaulet) [inline conf/visualization/default.yaml:2]: Please don't change the default, it will affect everyone

File: torch_points3d/datasets/generative/modelnet.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+from torch_points3d.datasets.base_dataset import BaseDataset
+from torch_points3d.metrics.generation_tracker import GenerationTracker
+from torch_points3d.utils.download import download_url
+
+import torch_points3d.core.data_transform as cT
+import torch_geometric.transforms as T
+
+
+class SampledModelNet(InMemoryDataset):
```
Comment (nicolas-chaulet) [inline torch_points3d/datasets/generative/modelnet.py:19]: Could you subclass the existing modelnet class we have for classification? A lot of the logic is already implemented there.

File: torch_points3d/datasets/generative/shapenet.py [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+from torch_geometric.io import read_txt_array
+import torch_geometric.transforms as T
+from torch_points3d.core.data_transform import SaveOriginalPosId
+from torch_points3d.metrics.generation_tracker import GenerationTracker
+from torch_points3d.datasets.base_dataset import BaseDataset, save_used_properties
+from torch_points3d.utils.download import download_url
+
+
+
+class ShapeNet(InMemoryDataset):
```
Comment (nicolas-chaulet) [inline torch_points3d/datasets/generative/shapenet.py:20]: Same comment, please subclass the existing ShapeNet class with generative specific details if any.

File: torch_points3d/models/generative/DGCN.py [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+        tmp = points - mu.repeat(1, 1, nump)    # Bx3xN - Bx3xN -> Bx3xN
+        tmp_transpose = tmp.transpose(1, 2)     # Bx3xN -> BxNx3
+        covariance = torch.bmm(tmp, tmp_transpose)
+        covariance = covariance / nump
+        return mu, covariance   # Bx3x1 Bx3x3
+
+    def get_neighbors(self, pt1, pt2):
+        #BxMx3 -> BMx3
+        pt1_batch = np.indices(pt1.shape[0:2])[0].reshape(-1)
+        pt1_batch = torch.from_numpy(pt1_batch).cuda()
```
Comment (nicolas-chaulet) [inline torch_points3d/models/generative/DGCN.py:73]: The code should run on cpu or gpu. Please replace the `.cuda` with `.to(self.device)`

File: torch_points3d/models/generative/DGCN.py [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+    def optimize_parameters(self, epoch, batch_size):
+        """Calculate losses, gradients, and update network weights; called in every training iteration"""
+        self._num_epochs = epoch
+        self._num_batches += 1
+        self._num_samples += batch_size
+
+        self.d_step(epoch=epoch)
+        self.g_step(epoch=epoch)
+
+    def d_step(self, *args, **kwargs):
```
Comment (nicolas-chaulet) [inline torch_points3d/models/generative/DGCN.py:217]: Any chance you could also provide a "forward" only mode for inference?

File: torch_points3d/metrics/generation_tracker.py [hunk 7]
```diff
@@ -0,0 +1,29 @@
+from typing import Dict, Any
+import torch
+import torchnet as tnt
+
+from torch_points3d.metrics.confusion_matrix import ConfusionMatrix
+from torch_points3d.metrics.base_tracker import BaseTracker, meter_value
+from torch_points3d.models import model_interface
+
+
+class GenerationTracker(BaseTracker):
```
Comment (tchaton) [inline torch_points3d/metrics/generation_tracker.py:10]: Could you had some metrics there as :
Jensen-Shannon Divergence (JSD), Minimum Matching Distance (MMD) and Coverage (COV), the earth mover’s distance (EMD), the chamfer distance (CD) and the 1-nearest neighbor accuracy (1-NNA) from the paper.

File: torch_points3d/models/generative/DGCN.py [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+import itertools
+import torch.nn as nn
+from torch_points3d.core.schedulers.lr_schedulers import instantiate_scheduler
+from torch_points3d.core.schedulers.bn_schedulers import instantiate_bn_scheduler
+from torch_points3d.core.common_modules.gathering import gather
+from torch_geometric.data import DataLoader, InMemoryDataset, extract_zip, Data
+
+log = logging.getLogger(__name__)
+
+class DGCN(BaseModel):
```
Comment (tchaton) [inline torch_points3d/models/generative/DGCN.py:23]: Could you make a BaseGanModel ?

File: torch_points3d/models/generative/DGCN.py [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+        like_var12 = self.chamfer_loss(var1,var2) / float(pt1_M)
+        # import pdb
+        # pdb.set_trace()
+
+
+        #print('mu: {} var: {}'.format(like_mu12.item(), like_var12.item())) 
+              
+        return like_mu12, like_var12
+
+    def get_pairs(self, data1, data2):
```
Comment (tchaton) [inline torch_points3d/models/generative/DGCN.py:148]: Do you use this function ?

File: torch_points3d/modules/DGCN/helpers.py [hunk 10]
```diff
@@ -0,0 +1,130 @@
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from torch.autograd import Variable
+# from pdb import set_trace as brk
+
+def get_edge_features_xyz(x, pc, k, num=-1):
+    """
```
Comment (tchaton) [inline torch_points3d/modules/DGCN/helpers.py:8]: Could you clean up a little bit the docstring and add some unit tests ?

File: torch_points3d/modules/DGCN/helpers.py [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+
+    # centralize
+    central = x.unsqueeze(3) # [B, d, N, 1]
+    central = central.repeat(1, 1, 1, k) # [B, d, N, k]
+
+    ee = torch.cat([central, neighbors-central], dim=1)
+    assert ee.shape == (B, 2*dims, N, k)
+    return ee
+
+class ChamferLoss(nn.Module):
```
Comment (tchaton) [inline torch_points3d/modules/DGCN/helpers.py:98]: Could you move the ChamferLoss to core/losses ?

File: torch_points3d/modules/DGCN/main_modules.py [hunk 12]
```diff
@@ -0,0 +1,99 @@
+import torch
```
Comment (tchaton) [inline torch_points3d/modules/DGCN/main_modules.py:1]: maybe rename gan_modules ?

File: torch_points3d/modules/DGCN/main_modules.py [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+from torch_points3d.modules.DGCN.modules import DeconvModule
+from torch_points3d.core.common_modules.dense_modules import MLP1D, Conv1D
+
+class PointGenerator(nn.Module):
+    def __init__(self, 
+        point_scales,
+        latent_space,
+        config,
+    ):
+        super(PointGenerator, self).__init__()
```
Comment (tchaton) [inline torch_points3d/modules/DGCN/main_modules.py:17]: Keep same arguments as within the configuration files.
And use _ as they are internal properties.
example: self.softmax -> self._use_softmax

File: conf/models/generative/dgcn.yaml [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+                [512, 256, 64],
+                [512, 256, 64],
+                [512, 256, 64],
+                [512, 256, 64],
+                # [256, 128, 64],
+                # [512, 256, 64],
+                # [1024, 256, 64],
+                # [2048, 512, 256, 64],
+            ]
+            classifier: PointNet2
```
Comment (tchaton) [inline conf/models/generative/dgcn.yaml:27]: I thought there were using PointNet within their paper for the discriminator :) We should add PointNet to the API so it implements the paper model

File: torch_points3d/modules/DGCN/modules.py [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+        **kwargs
+    ):
+        super(DeconvModule, self).__init__()
+
+        self.nn = nn.ModuleDict()
+
+        self.nn["maxpool"] = nn.MaxPool2d((1,maxpool),(1,1))
+        self.nn["upsample_cov"] = BilateralInterpolationModule(Fin, Fout, num_k//2, 1, softmax=softmax)
+        self.nn["bn_uc"] = nn.BatchNorm1d(Fout)
+        self.nn["relu_uc"] = nn.LeakyReLU(inplace=True)
```
Comment (tchaton) [inline torch_points3d/modules/DGCN/modules.py:30]: rename activation as you are using leakyRelu

File: torch_points3d/modules/DGCN/modules.py [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+class BilateralInterpolationModule(BaseModule):
+    """ Edge Convolution using 1x1 Conv h
+    [B, Fin, N] -> [B, Fout, N]
+    """
+    def __init__(self, Fin, Fout, k, num, softmax=True):
+        super(BilateralInterpolationModule, self).__init__()
+        self.k = k
+        self.Fin = Fin
+        self.Fout = Fout
+        self.softmax = softmax
```
Comment (tchaton) [inline torch_points3d/modules/DGCN/modules.py:88]: Same here for properties.

File: torch_points3d/datasets/generative/modelnet.py [hunk 17]
```diff
@@ -0,0 +1,183 @@
...
+    def len(self):
+        return super().len() #* len(self.scales)
+
+    def get(self, idx):
+        sample = super().get(idx) #// len(self.scales))
+
+        samples = []
+        data = Data()
+        for scale in self.scales:
+            transform = T.Compose([
```
Comment (tchaton) [inline torch_points3d/datasets/generative/modelnet.py:126]: I think we could implement a multi_scale fixed random sampling transform and add it the config files.
By doing so, you won't have to copy / paste both dataset.

File: torch_points3d/modules/DGCN/main_modules.py (hunk 13 above)
Comment (CCInc) [reply torch_points3d/modules/DGCN/main_modules.py]: I am not sure what you mean here, are these all internal properties?

File: conf/models/generative/dgcn.yaml (hunk 14 above)
Comment (CCInc) [reply conf/models/generative/dgcn.yaml:27]: Yes they did, but I thought using PointNet2 would be okay. But maybe it will be too powerful for the generator.

Comment (CCInc) [reply conf/models/generative/dgcn.yaml:27]: But in my testing the network didn't collapse so I think it should be fine. Maybe if someone else wants to implement PointNet to the API we can test it with that!

File: torch_points3d/metrics/model_checkpoint.py [hunk 18]
```diff
@@ -88,19 +92,26 @@ def load(checkpoint_dir: str, checkpoint_name: str, run_config: DictConfig, stri
...
-                try:
-                    optimizer_params = self.run_config.training.optim.optimizer.params
-                except:
-                    pass
-                optimizer = optimizer_cls(model.parameters(), **optimizer_params)
-                optimizer.load_state_dict(optimizer_config[1])
-                return optimizer
+                optimizers_out = []
+                if self.optimizer is not None:
+                    optimizers_config = [self.optimizer]
```
Comment (CCInc) [inline torch_points3d/metrics/model_checkpoint.py:100]: this might not be the best way to implement multiple optimizers
